### DISCUSSION TITLE
Update-2022-12-02

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#198] [Announcements]
Greetings everyone,

I have been busy the past few weeks fleshing out the GUI layer, implementing more widget code.  To be frank, the code is a bit of a hodge-podge, and will require revisiting at some point. I'm still not convinced whether we should implement all our own GUI code (as apparently Blender has done), or use an existing toolkit (a la Maya, Houdini) if we can manage a clean cross-platform FFI interface.

The big news is that @awolven has implemented the first pass at a Vulkan/Metal interface. I have got it running on my Mac, and will be looking at the code over the next little while. This is a huge step in bringing the system up to date and out of OpenGL 1.1 mode.

https://user-images.githubusercontent.com/103860793/205388611-f6c99c77-d4ae-4c46-a29f-46e4c7c1720e.mov

**Comment by jolby** [comment]
@kaveh808 That is looking great! I think it would be really cool to do our own GUI code using kons-9 to do the drawing. But I can also see the time saving practicality of using an existing toolkit.

As an aside, I've been hacking away at an experimental backend for McCLIM using SDL2/OpenGL/Skia. I'm still on my McCLIM backend training wheels, but if kons-9 does make its own GUI toolkit, I would love to try making a kons-9 based McCLIM backend.

**Comment by awolven** [comment]
There is no question in my mind that we can develop our own UI Toolkit which will be just as good or even better than what other major software products are using. We should look into making one from scratch but also we should look into CLIM.Sent from my Galaxy
-------- Original message --------From: Kaveh Kardan ***@***.***> Date: 12/2/22  4:23 PM  (GMT-05:00) To: kaveh808/kons-9 ***@***.***> Cc: Andrew Wolven ***@***.***>, Mention ***@***.***> Subject: [kaveh808/kons-9] Update-2022-12-02 (Discussion #198) 
Greetings everyone,
I have been busy the past few weeks fleshing out the GUI layer, implementing more widget code.  To be frank, the code is a bit of a hodge-podge, and will require revisiting at some point. I'm still not convinced whether we should implement all our own GUI code (as apparently Blender has done), or use an existing toolkit (a la Maya, Houdini) if we can manage a clean cross-platform FFI interface.
The big news is that @awolven has implemented the first pass at a Vulkan/Metal interface. I have got it running on my Mac, and will be looking at the code over the next little while. This is a huge step in bringing the system up to date and out of OpenGL 1.1 mode.
https://user-images.githubusercontent.com/103860793/205388611-f6c99c77-d4ae-4c46-a29f-46e4c7c1720e.mov

**Comment by kaveh808** [comment]
It is tempting to embark down the road to our own UI toolkit. We just have to be aware that it is a major undertaking with a lot of time-consuming and frankly tedious development. For example, just thinking of having to implement a file selection dialog from scratch is enough to make one shudder. :)

Someone (or someones) will need to seriously commit to this for a length of time.

**Comment by awolven** [comment]
Aiming low is the worst attitude we can have about this project because we would already have failed before we even start. Sent from my Galaxy
-------- Original message --------From: Kaveh Kardan ***@***.***> Date: 12/2/22  7:06 PM  (GMT-05:00) To: kaveh808/kons-9 ***@***.***> Cc: Andrew Wolven ***@***.***>, Mention ***@***.***> Subject: Re: [kaveh808/kons-9] Update-2022-12-02 (Discussion #198) 
It is tempting to embark down the road to our own UI toolkit. We just have to be aware that it is a major undertaking with a lot of time-consuming and frankly tedious development. For example, just thinking of having to implement a file selection dialog from scratch is enough to make one shudder. :)
Someone (or someones) will need to seriously commit to this for a length of time.

**Comment by ghost** [comment]


**Reply by kaveh808** [reply]
We have something similar in how our demo files work.  The difference is you C-c C-c to evaluate the forms. They don't get updated as you type. I prefer it that way, as I often don't want evaluation as I'm updating the code.
